### PR TITLE
Refactor subscriptions

### LINKED ISSUE #514: Refactor subscriptions to reduce code duplication
The current method for creating and maintaining subscriptions to lists like users, instances, saves, etc requires a lot of boilerplate code to set up. This should ideally be reduced with an abstraction layer that can handle most of the duplicated code for each type of subscription.

~~This abstraction layer should be able to handle showing and filtering elements in a table without loading all elements onto the client. For an example of this see the search mod feature in the Mod Pack view in the Web UI.~~ Edit: scrapped as this introduces a lot of complexity for little gain.

### PR DESCRIPTION
**Pull request by Hornwitser**
Refactors the way subscriptions work into being mappings that are synced across the network. To achieve this the events for subscribable resources have been standardised to contain a single `updates` property containing a list of updates values, where the values all have an `id`, `updatedAt` and `isDeleted` property. `id` is a number or string that uniquely identifies a given entry, `updatedAt` is a millisecond Unix timestamp of when the entry was last updated and `isDeleted` is a boolean indicating if the element has been deleted.  The EventSubscriber uses these properties to keep its copy of the resource in sync.

This implementation has one known issue with deleted entries being discarded by the controller. This breaks the resync logic which assumes sending a SubscriptionRequest with a `lastRequestTime` in the past gives updates with `isDeleted = true` for all entries that were deleted since that timestamp. Since this doesn't happen deleted entries stay in client lists that have been loaded but were not subscribed to at the time of deletion.

Resolves #514.

### CODE REVIEW

File: test/lib/subscriptions.js [hunk 1]
```diff
@@ -245,158 +180,144 @@ describe("lib/subscriptions", function() {
...
-			it("should not notify a link who unsubscribed from all all channels", async function() {
+			it("should not notify a link who unsubscribed", async function() {
 				const connectorData = connectorSetupDate[0];
 				const request = new lib.SubscriptionRequest(RegisteredEvent.name, false);
-				await subscriptions._handleRequest(request, connectorData.src, connectorData.dst);
-				mockController.wsServer.controlConnections.get(0).connector.sentMessages.pop(); // Response to request
+				await subscriptions.handleRequest(
+					getLink(connectorData.id), request, connectorData.src, connectorData.dst,
+				);
+				getLink(0).connector.sentMessages.pop(); // Response to request
```
Comment (Cooldude2606) [inline test/lib/subscriptions.js:195]: This line is no longer required. Originally it was because all subscription requests gave a responce, but this is no longer true. With the new code `sentMessages` is empty at this point, and `pop` is returning `undefined` so this is effectivly a nop. Test passes with line removed.

File: test/lib/subscriptions.js [hunk 2]
```diff
@@ -407,178 +328,135 @@ describe("lib/subscriptions", function() {
...
+				registeredEvent.subscribe((values, synced) => { calledWith = [values, synced]; });
+				const [msg] = await events.once(mockControl.connector, "send");
+				assert(!registeredEvent.synced, "synced was set before response");
+				mockControl.connector.emit("message", new lib.MessageResponse(0, msg.dst, msg.src));
+				await new Promise(resolve => setImmediate(resolve));
+				assert(calledWith !== undefined, "callback not called");
+				assert.deepEqual(calledWith, [[], true]);
+				assert(registeredEvent.synced, "synced was not set");
+			});
+			it("should be called con connector connect event", async function() {
```
Comment (Cooldude2606) [inline test/lib/subscriptions.js:440]: Minor spelling error: `con` -> `on`

**Review by Cooldude2606 — COMMENTED** [review]
Other comments:
- HostConnection contains 4 changes that could be covered by a single test, should be added to the todo list.
- InstanceInfo should have a round trip test.

File: packages/lib/src/data/messages_instance.ts [hunk 3]
```diff
@@ -251,60 +283,100 @@ export class InstanceStartRequest {
...
+		);
+	}
+
+	/** Concatenation of instance id and name forming a unique string for this save */
+	get id() {
+		return `${this.instanceId}/${this.name}`;
+	}
+
+	get updatedAt() {
+		return this.mtimeMs;
```
[the remark this answers is not in the document]
Comment (Hornwitser) [reply packages/lib/src/data/messages_instance.ts:326]: Good catch 👍, I have put together a fix for this.

File: packages/web_ui/src/components/InstancesPage.tsx [hunk 4]
```diff
@@ -96,7 +96,7 @@ export default function InstancesPage() {
 			</>}
 		/>
 
-		<InstanceList instances={instanceList} />
+		<InstanceList instances={instances} />
```
[the remark this answers is not in the document]
Comment (Hornwitser) [reply packages/web_ui/src/components/InstancesPage.tsx:99]: Yes this is correct. The other places feeds the data to the Table components which can't be changed to take a Map, but this component is our own and was changed to accept a Map.

File: packages/lib/src/data/ModPack.ts [hunk 5]
```diff
@@ -58,7 +58,7 @@ export default class ModPack {
 	/**
 	 * Id of this mod pack
 	 */
-	id?: number;
+	id: number = Math.random() * 2**31 | 0;
```
[the remark this answers is not in the document]
Comment (Hornwitser) [reply packages/lib/src/data/ModPack.ts:61]: I went one step further and made activeModPack undefined until syncMods is called.

File: packages/web_ui/src/components/ModPackViewPage.tsx [hunk 6]
```diff
@@ -245,7 +245,7 @@ type ModsTableProps = {
 };
 function ModsTable(props: ModsTableProps) {
 	const [showAddMods, setShowAddMods] = useState(false);
-	const [modList] = useModList();
+	const [modsMap] = useMods();
```
[the remark this answers is not in the document]
Comment (Hornwitser) [reply packages/web_ui/src/components/ModPackViewPage.tsx:248]: The issue is that the `mods` name is already in for the processed list of mods created on line 265.

File: packages/web_ui/src/components/RoleViewPage.tsx [hunk 7]
```diff
@@ -13,35 +13,27 @@ import PluginExtra from "./PluginExtra";
 import { notifyErrorHandler } from "../util/notify";
 
 
-export type RawRoleState = Partial<lib.Role> & {
-	loading?: boolean;
-	missing?: boolean;
-	present?: boolean;
-};
-function useRole(id: number): [RawRoleState, () => void] {
+function useRole(id: number): [Readonly<lib.Role> | undefined, boolean, () => void] {
```
[the remark this answers is not in the document]
Comment (Hornwitser) [reply packages/web_ui/src/components/RoleViewPage.tsx:16]: At the start of the project all the models were initially created in whatever file that needed them first and then moved over to the models directory when more than one place needed.

Thinking about it now, that kind of seems stupid. How would anyone else know to find useRoles in RoleViewPage... I'll move it if/when I introduce subscriptions for roles.

**Comment by Hornwitser** [comment]
My changes are ready for review now. The only issue not addressed is the changes which could be covered by a test in HostConnection because I don't really see any simple way to test the change. It's a corner case with a lot of dependencies.

**Review by Cooldude2606 — APPROVED** [review]
All my comments have been addressed.
